web/src/server/api/routers/scores.ts

```diff
@@ -364,6 +355,50 @@ export const scoresRouter = createTRPCRouter({
         },
       });
 
+      if (env.CLICKHOUSE_URL) {
+        const clickhouseTrace = await getTraceById(
+          input.traceId,
+          input.projectId,
+        );
+
+        if (!clickhouseTrace) {
+          // Fail silently while Postgres is in lead.
+          logger.error(
+            `No trace with id ${input.traceId} in project ${input.projectId} in Clickhouse`,
+          );
+        }
+
+        const clickhouseScore = await getScoreByNameAndTraceId(
+          input.projectId,
+          input.name,
+          input.traceId,
+          ScoreSource.ANNOTATION,
+        );
```

Comment:
Some additional context: a combined key of `projectId, name, traceId, observationId` OR `projectId, configId, traceId, observationId` can be used to satisfy uniqueness only on scores with source annotation. Right now we are not including observationId in this filter which might lead to unexpected behaviour. Let's take an example. Within a given project, we have traceA which a child generationA. traceA has score1 (name) set with source ANNOTATION. Now we want to set score1 (name) on generationA. This check would find the score1 attached to traceA and falsely flag this as a duplicate. So we'll need to include the observationId in the filter, and make sure the explicitly filter if it should be null (ie we're trying to set a trace score) or if the observation is present (ie we're trying to set an observation score). Does this make sense? :)

Reply:
@marliessophie Could you validate whether https://github.com/langfuse/langfuse/pull/4081/commits/f79bc50ef26cf543cc71ae04eb975d4271cabc3a covers the requirements that you've outlined?

My logic going forward:

- Search for projectId, name and (if provided) observationId
- _AND_ in addition for a match of either name _or_ config.

Since name is required according to the input object we shouldn't throw the error in the util function in any case, but receive an object if there is any match. In that case we reject the creation.